File: shimmer/api/rules.py
"""GET /api/rules — the rules the screens show, from the engine's catalog.

The browser kept its own copies of the Loudness choices, format ceilings,
EQ limits and more (docs/ARCHITECTURE.md §7). The screens read them from
here instead (docs/API.md §7), so the two can never disagree again.
"""
from __future__ import annotations

import dataclasses
from typing import Any, Dict

from fastapi import APIRouter

from ..core import catalog

router = APIRouter()


def rules() -> Dict[str, Any]:
    return {
        "loudness_targets": [dataclasses.asdict(t) for t in catalog.LOUDNESS_TARGETS],
        "default_loudness": catalog.DEFAULT_LOUDNESS,
        "formats": [dataclasses.asdict(f) for f in catalog.FORMATS],
        "default_format": catalog.DEFAULT_FORMAT,
        "cards": [dataclasses.asdict(c) for c in catalog.CARDS],
        "tools": list(catalog.TOOLS),
        "default_amount": catalog.DEFAULT_AMOUNT,
        "eq_limits": dict(catalog.EQ_LIMITS),
        "eq_types": list(catalog.EQ_TYPES),
    }


@router.get("/api/rules")
def get_rules() -> Dict[str, Any]:
    return rules()

File: static/js/rules.js
// rules.js — the engine's rules, fetched once from GET /api/rules.
//
// The screens used to keep their own copies of the Loudness choices, the
// format ceilings and more, and the copies drifted (docs/ARCHITECTURE.md
// §7). Every script reads them from here instead.

let rulesPromise = null;

/** The /api/rules answer, fetched once per page. */
export function loadRules() {
    if (!rulesPromise) rulesPromise = fetch('/api/rules').then((r) => r.json());
    return rulesPromise;
}

/** The LUFS level of a Loudness choice key, or null. */
export function loudnessLufs(rules, key) {
    const t = (rules.loudness_targets || []).find((x) => x.key === key);
    return t ? t.lufs : null;
}
